feat(celestia-adapter): update to celestia-node v0.11 and celestia app v1.0

zvolin opened the pull request:
# Description

This updates the `sov-celestia-adapter` to the latest (at the time of submitting) versions of celestia node (0.11.0-rc14) and app (1.0.0-rc17).

Sorry for such a big PR, but it was hard to introduce changes separately there

changes includes:
- a complete rewrite of celestia docker, this includes both `/docker` compose setup and the `Dockerfile` used in `demo-rollup/Makefile`. Now there is a single source of truth inside the `docker`. I tried to support all the existing use-cases, but please check if I haven't broken your workflows :pray: Also there seems to currently be no easy way to stop the running network in a way to resume it later, so it's recreated each time.
- usage of `celestia-types` in `sov-celestia-adapter`. `celestia-types` is a crate we maintain at @eigerco that consists of types compatible with celestia's protobuf and json API's. That way you shouldn't worry that you'll fall incompatible with the celestia network as we run integration tests for all those types with the real celestia-node. `celestia-types` have very little logic itself (yet) and is mostly about conforming to correct representations, so some types stayed as they were, but I tried to replace everything that was possible currently and we can then try to upstream the logic bit by bit.
- removal of `borsh` wherever `celestia-types` were included. We currently don't support borsh, but we can introduce the support for it in `celestia-types` wherever we don't include the `tendermint` types. All those places were currently marked as `TODO`
- recreation and addition of test data, updating all the tests and introducing a few new ones

## Linked Issues
- Fixes #461, fixes #480, 
- Related to # (issue) 

## Testing

Ran tests for sov-celestia-adapter
ran  demo-prover (with SKIP_PROVER=1) and demo-rollup + make test-create-token

## Docs

Updated the docs were relevant, also included some inline documentation

zvolin committed:
chore: create a new docker compose setup with local celestia network

zvolin committed:
feat: refactor celestia-adapter to use celestia-types

zvolin committed:
feature guard celestia-rpc crate

zvolin committed:
switch to address from celestia-types

zvolin committed:
cleanup, satisfy the clippy

zvolin committed:
switch to celestia_types commitment generation

zvolin committed:
Remove shares parsing

zvolin committed:
don't try to translate EDS to shares

zvolin committed:
treat extended data square as plain binary blobs, not shares

zvolin committed:
Simplify dah validation

zvolin committed:
Use namespaced shares for completeness proof

zvolin committed:
rebase fixes

zvolin committed:
Update test header.json

zvolin committed:
Update celestia-node and celestia-tendermint

zvolin committed:
Update test data for celestia adapter

zvolin committed:
fix existing tests

zvolin committed:
remove hex-literal dependency

zvolin committed:
Add tests for txs verification

zvolin committed:
Update lockfiles

zvolin committed:
fix linting

zvolin committed:
Add protoc to the test workflow

zvolin commented:
I gotta check that https://github.com/Sovereign-Labs/sovereign-sdk/actions/runs/6338399051/job/17219162315

zvolin committed:
Allow Zlib licenses in dependencies

zvolin commented:
@preston-evans98 so I allowed the [`Zlib`](https://www.zlib.net/zlib_license.html) licenses in the `deny.toml` as it seems minimal and very permissive. If you don't want to allow it for the sovereign sdk tho, I can just hardcode the constants that we build with `const_format` (the crate with zlib license) directly in `celestia-types`

preston-evans98 commented on .github/workflows/rust.yml:
```diff
@@ -145,6 +149,10 @@ jobs:
       - uses: actions/checkout@v3
         with:
           submodules: true
+      - name: Install Protoc
+        uses: arduino/setup-protoc@v2
+        with:
+          version: "23.2"
```
Where does the protoc compiler get introduced? Is this from `celestia-types`?

zvolin replied:
It's a dependency of our other pkg, `celestia-proto`.

zvolin replied:
It's used to build MsgPayForBlobs and other celestia and cosmos relevant protos

preston-evans98 replied:
Makes sense, thanks!

preston-evans98 commented on adapters/celestia/src/verifier/mod.rs:
```diff
@@ -182,20 +184,19 @@ impl da::DaVerifier for CelestiaVerifier {
         let mut tx_iter = txs.iter();
         let square_size = block_header.dah.row_roots.len();
-        for (blob, tx_proof) in namespace.blobs().zip(inclusion_proof.into_iter()) {
+        for (blob, tx_proof) in verified_shares.blobs().zip(inclusion_proof.into_iter()) {
```
This doesn't look like a new bug - but I think this check is broken. If `inclusion_proof` is an empty iterator, this loop will never execute. Looks like we need to check that `inclusion_proof.len()` == `verified_shares.blobs().len()`

zvolin replied:
this case should be covered now

zvolin replied:
https://github.com/Sovereign-Labs/sovereign-sdk/pull/948/commits/d755a2d3a2a6b15da77ea82abdb553da554ffcc0

preston-evans98 replied:
Fanstastic, thanks @zvolin!

preston-evans98 commented:
Thanks for the hard work here @zvolin! Generally the diff looks fine - I feel pretty confident that the changes here didn't introduce any new bugs. Unfortunately, I'm not at all confident that there aren't important bugs in the verifier (in fact, I think I spotted one fatal one during this review). 

Any updates from the Celestia side about incorporating `blob.sender` information into the data namespace directly? That would let us drastically simplify the logic and remove nearly all of the surface for bugs. If so, let's keep pushing! If not, we may need to consider a significant refactor or even a total rewrite of the verifier logic

zvolin committed:
Add protoc to the jobs where it was missing

zvolin committed:
Update demo-rollup readme with the changes to local setup

zvolin committed:
handle the case where amount of blobs is different than amount of proofs

zvolin commented:
> Any updates from the Celestia side about incorporating blob.sender information into the data namespace directly

Yes.

So I was initially searching for a way to just get relevant PFB's with proofs, so that the adapter doesn't need to fetch the entire EDS and recreate the proofs for only the PFB's it's interested in. The celestia team is aware of the issue that it's currently not possible to get the metadata relevant only for the blobs of the given namespace and will likely introduce some way to do it gently in a future, perhaps with an addition of a new node type, partial node (or namespaced node).

Quoting @Wondertan:

> FN and PN can theoretically index the reserved namespace to store that info and provide over API. If that becomes a very common usecase and we decide that core protocol should not handle that, then the node will likely provide this functionality.
>
> The node will need to optimize its BlobModule proofs. Proving the inclusion of a blob is equivalent to the inclusion of its PFB and the PFB inclusion proof is much smaller than the blob’s.
To achieve this FNs and PNs would need to parse the reserved namespace which contains all PFBs and metadata such as signer info. 
>
> Besides just described optimization, there is your use case which would benefit from node parsing and indexing PFBs for blobs. And in general, allowing  devs using our APIs to inspect PFBs is a huge enabler for various dashboards and explorers

I asked about the inclusion of the `sender` (or `signer`) metadata directly into blob shares and the chances are rather slim, again quoting @Wondertan:

> That's a simpler but consensus breaking solution and why I involved app team. My problem is that does not scale for other fields inside of PFB tx and duplicates data. We don't wanna break consensus everytime someone needs to map blob to some metadata in PFB and abstracting away this (necessary) separation between Blobs and PFBs on the node's API seems like a better solution to me.

preston-evans98 commented:
> Quoting @Wondertan:

Thanks for the update! Unfortunately, I don't @Wondertan has the full picture on the use case here (changing node APIs doesn't help in this case, since we need to cryptographic verification). I'll ping the Celestia folks and see if we can get discussions started again

zvolin commented:
> changing node APIs doesn't help in this case, since we need to cryptographic verification

I was asking for a way to obtain the relevant pfb's with inclusion proofs just for them. Currently when we fetch the whole PFB namespace we get all the pfbs for block with proof for all the pfbs even tho we care only about the few. The adapter currently fetches the entire EDS and recreates the proofs only for relevant pfbs and that was a problem I was asking to address, thus the node's api change.

zvolin commented:
Including the sender directly in blob's shares is kind of related but also separated topic and only the last quote is in regards to that

preston-evans98 commented:
Got it, thanks for the additional context

zvolin committed:
fix long running make start on the CI

zvolin committed:
align prover benchmarks

zvolin committed:
remove fuzzing namespace group creation from base64

zvolin commented:
rebased and fixed stuff from previous CI run

zvolin committed:
revert celestia-address changes from: Initialize genesis state form genesis files. (#972)

zvolin commented:
rebased one more time. aligned new method name in tests from #977 and reverted changes to celestia addresses from #972. Could you retrigger one more time please @preston-evans98 ?

zvolin commented:
rebased

zvolin commented:
Closing in favor of #997
